apm-agent-attach-cli/src/test/java/co/elastic/apm/attach/DiscoveryRulesTest.java

```diff
@@ -75,6 +75,16 @@ void testIncludeMultiple() {
         assertThat(discoveryRules.isMatching(jvmWithPid("3"), userRegistry)).isFalse();
     }
 
+    @Test
+    void testDiscoveryRequired() {
+        assertThat(discoveryRules.isDiscoveryRequired()).isFalse();
+        discoveryRules.includePid("1");
+        discoveryRules.includePid("2");
+        assertThat(discoveryRules.isDiscoveryRequired()).isFalse();
+        discoveryRules.excludePid("3");
+        assertThat(discoveryRules.isDiscoveryRequired()).isTrue();
```

Comment:
The way I understand this is when we include PIDs, we assume that the PID was provided from somewhere, and thus we don't have to attempt discoverey. However, I'm not sure to get why we shouldn't do the same for excluding PIDs.

Reply:
Exclude PID doesn't give any advice as to what process to attach. It doesn't make sense to provide it without additional discovery rules, which means we need to do discovery.

Reply:
Note that even if you use `--include-pid` with any additional discovery rule, it will be executed after discovery and in the order defined by the configuration.
Don't treat this combination of rules as such that make sense in reality, they only unit-test the logic

Reply:
I see, we only allow to take the shortcut when including a specific PID because in this case we know other discovery rule is necessary.

Reply:
More accurately: we only _**can**_ take the shortcut when including _**only**_ specific PIDs because in this case discovery is not required _**and we have enough info to attach**_